File: problemtools/exoplanetaryroverslam/submissions/partially_accepted/naive_verify.cpp
#include <iostream>
#include <vector>

using namespace std;

const long long MOD = 10007;
const long long A_LCG = 911382323;
const long long C_LCG = 972663749;

vector<vector<long long>> generate_matrix(int n, long long seed) {
    vector<vector<long long>> matrix(n, vector<long long>(n));

    long long x = seed % MOD;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            x = (A_LCG * x + C_LCG) % MOD;
            matrix[i][j] = x;
        }
    }

    return matrix;
}

vector<vector<long long>> multiply_matrices(
    const vector<vector<long long>>& A,
    const vector<vector<long long>>& B,
    int n
) {
    vector<vector<long long>> C(n, vector<long long>(n, 0));

    for (int i = 0; i < n; ++i) {
        for (int k = 0; k < n; ++k) {
            long long aik = A[i][k];
            if (aik == 0) continue;

            for (int j = 0; j < n; ++j) {
                C[i][j] = (C[i][j] + aik * B[k][j]) % MOD;
            }
        }
    }

    return C;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    if (!(cin >> N >> M)) return 0;

    vector<long long> seeds(M);
    for (int i = 0; i < M; ++i) {
        cin >> seeds[i];
    }

    vector<vector<long long>> E(N, vector<long long>(N));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cin >> E[i][j];
        }
    }

    vector<vector<long long>> P = generate_matrix(N, seeds[0]);

    for (int i = 1; i < M; ++i) {
        vector<vector<long long>> mat = generate_matrix(N, seeds[i]);
        P = multiply_matrices(P, mat, N);
    }

    cout << (P == E ? "YES\n" : "NO\n");

    return 0;
}

File: problemtools/exoplanetaryroverslam/submissions/partially_accepted/naive_verify.py
import sys

MOD = 1000000007
A_LCG = 911382323
C_LCG = 972663749

def generate_matrix(n, seed):
    x = seed % MOD
    matrix = [[0] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            x = (A_LCG * x + C_LCG) % MOD
            matrix[i][j] = x

    return matrix

def multiply_matrices(A, B, n):
    result = [[0] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            aij = A[i][j]
            if aij == 0:
                continue
            for k in range(n):
                result[i][k] = (result[i][k] + aij * B[j][k]) % MOD

    return result

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    ptr = 0
    n = int(input_data[ptr])
    ptr += 1

    m = int(input_data[ptr])
    ptr += 1

    seeds = [int(input_data[ptr + i]) for i in range(m)]
    ptr += m

    E = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            E[i][j] = int(input_data[ptr])
            ptr += 1

    prod = generate_matrix(n, seeds[0])

    for seed in seeds[1:]:
        mat = generate_matrix(n, seed)
        prod = multiply_matrices(prod, mat, n)

    print("YES" if prod == E else "NO")

if __name__ == "__main__":
    solve()